x-pack/metricbeat/module/statsd/server/data.go

```diff
@@ -0,0 +1,191 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package server
+
+import (
+	"bytes"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/helper/server"
+)
+
+var errInvalidPacket = errors.New("invalid statsd packet")
+
+type metricProcessor struct {
+	registry      *registry
+	reservoirSize int
+}
+
+type statsdMetric struct {
+	name       string
+	metricType string
+	sampleRate string
+	value      string
+	tags       map[string]string
+}
+
+func splitTags(rawTags []byte, kvSep []byte) map[string]string {
+	tags := map[string]string{}
+	for _, kv := range bytes.Split(rawTags, []byte(",")) {
+		kvSplit := bytes.SplitN(kv, kvSep, 2)
+		if len(kvSplit) != 2 {
+			continue
+		}
+		tags[string(kvSplit[0])] = string(kvSplit[1])
+	}
+	return tags
+}
+
+func parseSingle(b []byte) (statsdMetric, error) {
+	// format: <metric name>:<value>|<type>[|@samplerate][|#<k>:<v>,<k>:<v>]
+	// alternative: <metric name>[,<k>=<v>,<k>=<v>]:<value>|<type>[|@samplerate]
```

Comment:
maybe simpler. I went with this because i started without supporting tags. I still think there's a chance this is faster, however it probably doesn't matter too much. I suggest we get this in as-is, unless you think i should change it now?

Reply:
Let's go on as is :+1:  I don't think it should be changed now.